[ios] Link PlatformView back to semantics tree

cyanglaz opened the pull request:
The PlatformView does not have a semantics container when added to semantics tree, this PR gives it a semantics container to ensure accessibility traversal works.

This fixes https://github.com/flutter/flutter/issues/135504, which is a regression of https://github.com/flutter/engine/commit/b8332e3c4870512a941d685d7d375b1d2d62983c

Before https://github.com/flutter/engine/commit/b8332e3c4870512a941d685d7d375b1d2d62983c, the traversal works because the PlatformView is added to the accessibilityElements of the FlutterPlatFormViewSemanticsContainer, which implicitly made the FlutterPlatFormViewSemanticsContainer as the PlatformVIew's AccessibilityContainer. 

Now we use the PlatformVIew as the nativeAccessibility of the  FlutterPlatFormViewSemanticsContainer, we need to expilicitly set the container. 

This needs to be cherry-picked since the commit caused the regression was cherry-picked in https://github.com/flutter/engine/commit/b8332e3c4870512a941d685d7d375b1d2d62983c

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

cyanglaz commented on shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:
```diff
@@ -882,12 +880,6 @@ - (id)nativeAccessibility {
 
-#pragma mark - UIAccessibilityContainer overrides
-
-- (NSArray*)accessibilityElements {
-  return @[ _platformView ];
-}
```
We implemented nativeAccessibility, this is not necessary.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h:
```diff
@@ -234,6 +235,8 @@ class FlutterPlatformViewsController {
   // returns nil.
   UIView* GetPlatformViewByID(int64_t view_id);
 
+  FlutterTouchInterceptingView* GetFlutterTouchInterceptingViewByID(int64_t);
```
Needs a documentation comment.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:
```diff
@@ -957,6 +962,7 @@ @implementation FlutterTouchInterceptingView {
   fml::scoped_nsobject<DelayingGestureRecognizer> _delayingRecognizer;
   FlutterPlatformViewGestureRecognizersBlockingPolicy _blockingPolicy;
   UIView* _embeddedView;
+  NSObject* _flutterAccessibilityContainer;
```
Needs a comment.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:
```diff
@@ -1035,6 +1041,14 @@ - (void)touchesCancelled:(NSSet<UITouch*>*)touches withEvent:(UIEvent*)event {
 - (void)touchesEnded:(NSSet*)touches withEvent:(UIEvent*)event {
 }
 
+- (void)setFlutterAccessibilityContainer:(NSObject*)flutterAccessibilityContainer {
```
Why is this being declared as an ivar and manual getter/setter methods, rather than declared as a property, with synthesized methods?

cyanglaz replied:
We don't need the getter, we only need the setter so the "hidden" getter, accessibilityContainer used by UIAccessibility can return the correct value.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:
```diff
@@ -19,6 +19,10 @@
 const float kFloatCompareEpsilon = 0.001;
 
+@interface FlutterTouchInterceptingView (Test)
+- (id)accessibilityContainer;
```
Please make a `_Test.h` file instead of re-declaring methods directly in test files.

stuartmorgan replied:
This entire redeclaration block should be removed in favor of including the test header.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/SemanticsObject.h:
```diff
@@ -19,6 +19,14 @@ constexpr float kScrollExtentMaxForInf = 1000;
 
+@protocol FlutterSemanticsProtocol <NSObject>
+// Adding functionalities for any object to assign accessibilityContainer.
+// For example, this allows the PlatformView be able to link back to the accessibility tree.
```
I'm confused by what this property is actually for. I see where it's set, but I don't see anything reading it except in tests. Am I missing some non-test behavior that's influenced by this?

cyanglaz replied:
It is read in `FlutterTouchInterceptingView`: accessibilityContainer, it is a UIAccessibilty property that is read by the UIKit.

This protocol was created so that we can call this method to `FlutterTouchInterceptingView`. I initially added the method to the `FlutterTouchInterceptingView` so that `FlutterTouchInterceptingView` can access the container in its `accessibiltiyContainer` method.  

Then I discovered that SemanticsObjectTests is ARC, so we couldn't not include `FlutterTouchInterceptingView` (which uses MRC features) in the header of `SemanticsObjectTests`

This does make the code more confusing.

cyanglaz commented:
@stuartmorgan I removed the protocol to reduce confusion, I also refactored the platform view related tests to a separate file + other review comments. PTAL

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterTouchInterceptingView+Tests.h:
`..._Test.h` seems to be the established convention in this directory.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h:
```diff
@@ -424,6 +432,10 @@ class FlutterPlatformViewsController {
 // Get embedded view
 - (UIView*)embeddedView;
+
+// The FlutterTouchInterceptingView assigns `flutterAccessibilityContainer` as its
```
Instance methods don't need to say that they are about the class they are on; this can just say "Sets `flutterAccessibilityContainer` as this view's `accessibilityContainer`."

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/SemanticsObjectTest_mrc.mm:
```diff
@@ -0,0 +1,89 @@
+  // Check if there's no more strong references to `platformView` after container and platformView
+  // are released.
+  [container release];
+  XCTAssertEqual(platformView.retainCount, 1u);
```
This is fragile to implementation details of the code using the platform view. It would help to at least wrap all the code between the declaration and here in an autorelease pool.

cyanglaz commented:
@stuartmorgan Updated, PTAL!

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h:
```diff
@@ -230,10 +231,17 @@ class FlutterPlatformViewsController {
+  // If the `FlutterPlatformViewsController` does not contain any `FlutterPlatformView` object or
+  // a `FlutterPlatformView` object associated with the view_id cannot be found, the method
+  // returns nil.
+  FlutterTouchInterceptingView* GetFlutterTouchInterceptingViewByID(int64_t);
```
Nit: I missed that this parameter doesn't have a name; it should be `int64_t view_id`.

stuartmorgan reviewed (approved):
LGTM with one last nit.
